activerecord/lib/active_record/autosave_association.rb

```diff
@@ -429,36 +434,44 @@ def save_collection_association(reflection)
       def save_has_one_association(reflection)
         association = association_instance_get(reflection.name)
         record      = association && association.load_target
+        return unless record && !record.destroyed?
 
-        if record && !record.destroyed?
-          autosave = reflection.options[:autosave]
+        inverse_belongs_to_association = inverse_belongs_to_association_for(reflection, record)
+        return if inverse_belongs_to_association && inverse_belongs_to_association.updated?
 
-          if autosave && record.marked_for_destruction?
-            record.destroy
-          elsif autosave != false
-            primary_key = Array(compute_primary_key(reflection, self)).map(&:to_s)
-            primary_key_value = primary_key.map { |key| _read_attribute(key) }
+        autosave = reflection.options[:autosave]
 
-            if (autosave && record.changed_for_autosave?) || _record_changed?(reflection, record, primary_key_value)
-              unless reflection.through_reflection
-                foreign_key = Array(reflection.foreign_key)
-                primary_key_foreign_key_pairs = primary_key.zip(foreign_key)
+        if autosave && record.marked_for_destruction?
+          record.destroy
+        elsif autosave != false
+          primary_key = Array(compute_primary_key(reflection, self)).map(&:to_s)
+          primary_key_value = primary_key.map { |key| _read_attribute(key) }
 
-                primary_key_foreign_key_pairs.each do |primary_key, foreign_key|
-                  association_id = _read_attribute(primary_key)
-                  record[foreign_key] = association_id unless record[foreign_key] == association_id
-                end
-                association.set_inverse_instance(record)
-              end
+          if (autosave && record.changed_for_autosave?) || _record_changed?(reflection, record, primary_key_value)
+            unless reflection.through_reflection
+              foreign_key = Array(reflection.foreign_key)
+              primary_key_foreign_key_pairs = primary_key.zip(foreign_key)
 
-              saved = record.save(validate: !autosave)
-              raise ActiveRecord::Rollback if !saved && autosave
-              saved
+              primary_key_foreign_key_pairs.each do |primary_key, foreign_key|
+                association_id = _read_attribute(primary_key)
+                record[foreign_key] = association_id unless record[foreign_key] == association_id
+              end
+              association.set_inverse_instance(record)
             end
+
+            saved = record.save(validate: !autosave)
+            raise ActiveRecord::Rollback if !saved && autosave
+            saved
           end
         end
       end
 
+      def inverse_belongs_to_association_for(reflection, record)
```

Comment:
This isn't meant as public API right? So It would be best to move it in the private section

Reply:
I think it's already private. All these methods are under [this access modifier](https://github.com/rails/rails/blob/78ce2994b23b4b68db07a910df252284f1d18954/activerecord/lib/active_record/autosave_association.rb#L277)

Reply:
Ah indeed. I got confused because the methods around are documented, which is mostly for public methods.